File: ETour-DotNet/Pricing/FareBandPolicy.cs
using ETour.Api.Options;
using Microsoft.Extensions.Options;

namespace ETour.Api.Pricing;

public class FareBandPolicy
{
    public int ChildMaxAge { get; }

    public FareBandPolicy(IOptions<PricingOptions> options)
    {
        ChildMaxAge = options.Value.ChildMaxAge;
    }

    public int CalculateAgeAtDeparture(DateOnly? birthDate, DateOnly? departureDate)
    {
        if (birthDate is null)
        {
            throw new ArgumentException("birthDate is required for every passenger");
        }
        if (departureDate is null)
        {
            throw new ArgumentException("departure date is required to price a booking");
        }
        if (birthDate > departureDate)
        {
            throw new ArgumentException(
                $"birthDate {birthDate:yyyy-MM-dd} is after the departure date {departureDate:yyyy-MM-dd}");
        }

        var b = birthDate.Value;
        var d = departureDate.Value;

        var age = d.Year - b.Year;
        if (d < b.AddYears(age)) age--;
        return age;
    }

    public bool IsChild(int age) => age <= ChildMaxAge;

    public FareBand Resolve(int age, bool withBed, string occupancy)
    {
        if (IsChild(age))
        {
            return withBed ? FareBand.CHILD_WITH_BED : FareBand.CHILD_WITHOUT_BED;
        }

        var requested = FareBandExtensions.FromOccupancy(occupancy);
        if (requested == FareBand.SINGLE || requested == FareBand.EXTRA_PERSON)
        {
            return requested.Value;
        }
        return FareBand.TWIN_SHARING;
    }
}

File: E-Tour/src/main/java/com/etour/pricing/FareBandPolicy.java
package com.etour.pricing;

import java.time.LocalDate;
import java.time.Period;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class FareBandPolicy {
    private final int childMaxAge;

    public FareBandPolicy(@Value("${etour.pricing.child-max-age:12}") int childMaxAge) {
        this.childMaxAge = childMaxAge;
    }

    public int calculateAgeAtDeparture(LocalDate birthDate, LocalDate departureDate) {
        if (birthDate == null) {
            throw new IllegalArgumentException("birthDate is required for every passenger");
        }
        if (departureDate == null) {
            throw new IllegalArgumentException("departure date is required to price a booking");
        }
        if (birthDate.isAfter(departureDate)) {
            throw new IllegalArgumentException(
                    "birthDate " + birthDate + " is after the departure date " + departureDate);
        }
        return Period.between(birthDate, departureDate).getYears();
    }

    public boolean isChild(int age) {
        return age <= childMaxAge;
    }

    public FareBand resolve(int age, boolean withBed, String string) {
        if (isChild(age)) {
            return withBed ? FareBand.CHILD_WITH_BED : FareBand.CHILD_WITHOUT_BED;
        }
        FareBand requested = FareBand.fromOccupancy(string);
        if (requested == FareBand.SINGLE || requested == FareBand.EXTRA_PERSON) {
            return requested;
        }
        return FareBand.TWIN_SHARING;
    }

    public int getChildMaxAge() {
        return childMaxAge;
    }
}
